### ISSUE TITLE
[FEATURE] Large Transition Batch Size

### ISSUE THREAD

**Issue by cjfcsjt** [iffiX/machin#24]
**Is your feature request related to a problem? Please describe.**
1. Transition batch size

I read the tutorial about RL in [spiningup](https://spinningup.openai.com/en/latest/algorithms/ppo.html). I found that for on-policy RL, they have a step to collect a set of trajectories in their pseudocode. However, in your documentation [Data flow in Machin](https://machin.readthedocs.io/en/latest/tutorials/data_flow_in_machin.html), you point out that 
> Currently, the constructor of the default transition implementation Transition requires batch size to be 1

and 

> Buffer.store_episode(): If you pass in a dict type transition object, it will be automatically converted to Transition 

In your PPO [examples/ppo.py](https://github.com/iffiX/machin/blob/master/examples/framework_examples/ppo.py) it seems that you only save one trajectory per iteration and update it. What should I do if I want to save a set of trajectories? Will such a change affect the update() part?

2. Multiple trajectories with one reward

In general scenarios, one trajectory (episode) will have one total reward. However, I encountered a case where multiple trajectories with only one total reward. For example:
> (trajectory1: [s,a,0,s,a,0,...,s,a], trajectory2 [s,a,0,s,a,0,...,s,a], trajectory3[s,a,0,s,a,0,...,s,a] ) ---> final reward 
> Imagine that many football players are playing the same game. They receive the same reward only when goal, 
> Imagine that generating a batch of noise sequences to attack the neural network will get only one reward which indicates the degraded performance of the neural network.

I give two solutions to this problem in the next section, but I am not sure which one is better. Could you give me some advice?

**Describe the solution you'd like**
For feature 1, it may realized as below(I am not sure whether this affect the update() part):
```python
while episode< max_episodes:
    for i in range(sub_episode_size) # add a loop here
        episode += 1
        tmp_observations = []
        while not terminal and step <= max_steps:
               # ....
               tmp_observations.append(...) # store transition
         ppo.store_episode(tmp_observations)
     ppo.update()
     # clean buffer
```
For feature 2, it may have two solutions. One is to assign the final reward to every trajectory:
```python
from collection import defaultdict
while episode< max_episodes:
    batch_trajectory = defaultdict[List]
    episode += batch_size
    while not terminal and step <= max_steps:
           # using batch state to generate batch action
           reward = env.step(batch_action)
           for i in range(batch_size):
               batch_trajectory[i] += batch_state[i] + batch_action[i] + reward # assign the same reward to each trajectory

     for i in range(batch_size):
         ppo.store_episode(batch_trajectory[i])
     ppo.update()
     # clean buffer
```
**Describe alternatives you've considered**

For feature 2, another solution may resort to the multi-agent RL. Each agent manages one trajectory and they receive the same reward from the environment. I found that Machin has a multi-agent algorithm implementation called MADDPG. From [spiningup](https://spinningup.openai.com/en/latest/algorithms/ppo.html) I found that this algorithm is only for continuous action space. Is there any plan to implement other multi-agent RL algorithms such as multi-agent PPO for discrete action space?
 

**Additional context**

**Comment by iffiX** [comment]
Hi,

---
### Answer to problem 1
For the first problem, your solution is the desired way to work with PPO as only the `update()` call will clear the buffer, you can store as many episodes as you like, or as long an episode as you wish if your buffer and memory capacity permits.

> Currently, the constructor of the default transition implementation Transition requires batch size to be 1

This design is used to ensure that the tensors in your Transition object representing a single step will be concatenatable.

### Answer to problem 2
The answer is complex, first, for the two examples given in your description:
> Imagine that many football players are playing the same game. They receive the same reward only when goal.
- If your football players can communicate with each other and share a global view
  - The best design is using a central controller to issue commands for all players simultaneously based on views collected from all players. In this way, you are dealing with a single agent MDP problem, and will not be troubled by problems such as reward oscillation caused by lack of synchronization in distributed POMDP problems. Some multi-agent algorithms can also be used in this scenario, such as MADDPG, since if you want to add new players, the single-agent model may suffer from expandability problems, while multi-agent algorithms do not.
- If your football players only have local views, and has a bandwidth restricted communication channel
  - The optimal solution is using multi-agent algorithms. These types of RL algorithms may focus on optimizing communication costs, allocate rewards to agents based on their contribution, etc, along with maximizing your target reward.
- If your football players only have local views
  - In this scenario, you can only use single-agent models on each agent, as there is no additional information to coordinate them.

> Imagine that generating a batch of noise sequences to attack the neural network will get only one reward which indicates the degraded performance of the neural network.
Could you please elaborate on how the "sequence" will attack the neural network? I haven't done much research on neural adversaries based on procedural processes.

The solution really depends on how you model your problem and the granularity of control, If you could describe your specific problem in detail I could help you with the design. Also, for MADDPG it can be tweaked to output discrete actions by outputting a softmaxed vector (continuous) and then sample a discrete action from a categorical distribution parameterized by the vector, however, this may not be the best solution for your exact need.

**Comment by cjfcsjt** [comment]
Many thanks for your kind reply!!
> Could you please elaborate on how the "sequence" will attack the neural network? I haven't done much research on neural adversaries based on procedural processes.

Let's say there is an agent which aims to attack a graph. The graph node is represented by an int type id. Then, the agent will produce some sequences which denote the edges in a graph. For example, the agent could generate a fixed-length sequence [X, 1, 2, 3, 4, 5, 6], which denotes 6 edges: set{ (X,1), (X, 2), (X,3), ..., (X,6) }, where X is a specific node. These six edges will inject into a clean graph to get a polluted graph, and finally, the reward will be a specific metric reported by a graph neural network trained on such a polluted graph. 

The above can be modeled as a MDP:
> s0 = [X,1], a0 = [2], r0 = 0,       
> s1 = [X,1, 2], a1 = [3], r1= 0,        
>  s2 = [X,1, 2, 3], a2 = [4], r2 = 0
> s3 = [X,1, 2, 3,4], a2 = [5], r3 = 0
> s4 = [X,1, 2, 3, ... ,5], a4 = [6], r4 = "real value metric" 
> s5 = [X,1, 2, 3, ... ,6] finished

The thing is that the agent have the ability to generate multiple sequences (e.g. three sequences)：
> [X, 1, 2, 3, 4, 5, 6], [Y, 4, 7, 2, 1, 7, 10], [Z, 6, 42, 20, 9, 12, 54] ( 3*6 = 18 edges here )


After injecting these three sequences into the clean graph, we get only a single reward. Since I am a novice in RL, I am not sure how to deal with such a case. One way to solve this is to make the reward shared with all three sequences, treat the problem as a single MDP, and use the on-policy algorithms to update based on these three sequences. Another way to solve this is that treat each sequence as an independent agent (I am not sure whether such a treatment is in accord with the multi-agent RL scenerios). The latter seems more complex. Could you give me some advice on these two solutions?

**Comment by iffiX** [comment]
According to your description, I will recommend you to model this problem as a single-agent MDP as there is no need to introduce additional complexity. I would recommend the following methods:
1. If the generated sequence is always fixed-length, why not view the whole sequence as one action? For three sequences, there would be 3 actions in an episode, with one reward. Or you can view them as 18 actions chained together [(X, 1), (X, 2),..., (Y, 4), (Y, 7)...(Z, 54)]. Either method will require you to deal with the sparse rewards problem in RL, which could be solved by reward shaping explained [here](https://medium.com/@m.k.daaboul/dealing-with-sparse-reward-environments-38c0489c844d), since there is no temporal dependence between these actions (i.e they actually happen at the same time), then you can assign the same reward to all actions, though I recommend you divide reward by the number of actions as a normalization since it is probably an implicit desire to let the model use the least number of actions to get the highest reward.
2. Consider using RNN networks as your critic network (PPO, usually on policy) or Q network(DQN, DDPG, usually off policy), your actor can generate these actions & sequences sequentially, but since their cumulative effect is evaluated only once, it is reasonable to consider them together as a variable-length action, like:
   > s = [X, Y, Z] 
   > a = [[1, 2, 3, 4, 5, 6], [4, 7, 2, 1, 7, 10], [6, 42, 20, 9, 12, 54]]
   > r = "real value metric"

   then you need to use the critic or Q network to evaluate the value V(s) or Q(s, a) using these variable length states and actions, you can to refer to this issue: 
   https://github.com/iffiX/machin/issues/8 
   for some help on that.
